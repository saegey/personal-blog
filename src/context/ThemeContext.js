import React from "react"

const defaultState = {
  dark: false,
  toggleDark: () => {},
}

const ThemeContext = React.createContext(defaultState)

// Getting dark mode information from OS!
// You need macOS Mojave + Safari Technology Preview Release 68 to test this currently.
const supportsDarkMode = () =>
  window.matchMedia("(prefers-color-scheme: dark)").matches === true

class ThemeProvider extends React.Component {
  state = {
    unitOfMeasure: false,
  }

  toggleUnit = () => {
    // console.log("oh hey")
    let unitOfMeasure
    if (this.state.unitOfMeasure === "imperial") {
      unitOfMeasure = "metric"
    } else {
      unitOfMeasure = "imperial"
    }
    localStorage.setItem("unitOfMeasure", unitOfMeasure)
    this.setState({ unitOfMeasure })
  }

  componentDidMount() {
    // Getting dark mode value from localStorage!
    const lsDark = localStorage.getItem("unitOfMeasure")
    if (lsDark) {
      this.setState({ unitOfMeasure: lsDark })
    } else if (supportsDarkMode()) {
      this.setState({ unitOfMeasure: "imperial" })
    }
  }

  render() {
    const { children } = this.props
    const { unitOfMeasure } = this.state
    return (
      <ThemeContext.Provider
        value={{
          unitOfMeasure,
          toggleUnit: this.toggleUnit,
        }}
      >
        {children}
      </ThemeContext.Provider>
    )
  }
}

export default ThemeContext

export { ThemeProvider }
