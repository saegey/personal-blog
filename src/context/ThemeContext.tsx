import { createContext, Component } from 'react'

const defaultState = {
  unitOfMeasure: 'imperial',
  toggleUnit: () => {},
}

const ThemeContext = createContext(defaultState)

class ThemeProvider extends Component {
  state = {
    unitOfMeasure: 'imperial',
  }

  toggleUnit = () => {
    console.log('unit toggle')
    let unitOfMeasure
    if (this.state.unitOfMeasure === 'imperial') {
      unitOfMeasure = 'metric'
    } else {
      unitOfMeasure = 'imperial'
    }
    localStorage.setItem('unitOfMeasure', unitOfMeasure)
    this.setState({ unitOfMeasure })
  }

  componentDidMount() {
    const unit = localStorage.getItem('unitOfMeasure')
    if (unit) {
      this.setState({ unitOfMeasure: unit })
    } else {
      this.setState({ unitOfMeasure: 'imperial' })
    }
  }

  render() {
    const { children } = this.props
    const { unitOfMeasure } = this.state
    return (
      <ThemeContext.Provider
        value={{
          unitOfMeasure,
          toggleUnit: this.toggleUnit,
        }}
      >
        {children}
      </ThemeContext.Provider>
    )
  }
}

export default ThemeContext

export { ThemeProvider }
