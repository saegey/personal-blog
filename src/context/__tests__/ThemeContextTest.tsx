/**
 * @jest-environment jsdom
 */
import renderer from 'react-test-renderer'

import { Button } from 'theme-ui'
import { render, screen, cleanup, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

// import { ThemeProvider } from 'theme-ui'

import ThemeContext from '../ThemeContext'

import { theme } from './__test-utils__'

afterEach(cleanup)

describe('ThemeContext', () => {
  it('renders', () => {
    const json = renderer
      .create(
        <ThemeContext.Consumer>
          {theme => <h1>{theme.unitOfMeasure}</h1>}
        </ThemeContext.Consumer>
      )
      .toJSON()
    expect(json).toMatchSnapshot()
  })

  it('changes unit of it is toggled', async () => {
    render(
      <ThemeContext.Consumer>
        {theme => {
          return (
            <div>
              <Button title="Test Button" onClick={theme.toggleUnit} />
              <h1>{theme.unitOfMeasure}</h1>
            </div>
          )
        }}
      </ThemeContext.Consumer>
    )

			userEvent.click(screen.getByTitle('Test Button'))
    const blah = await screen.findByText('metric')
		expect(blah).toEqual('')



    // expect(json).toMatchSnapshot()
  })
})
