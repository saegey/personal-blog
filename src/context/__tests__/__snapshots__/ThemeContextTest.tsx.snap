// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeContext renders 1`] = `
<h1>
  imperial
</h1>
`;
